### PR TITLE
loops-programs_with_repetition-hard-q7

### PR DESCRIPTION
**Pull request by SamamthaH**
## PR Checklist
- [ ]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [ ]  Are the assertion fail messages good?
    - [ ]  Logical (does it make sense?)
    - [ ]  Readable (can a student understand what’s wrong?)
    - [ ]  Not too feed-y (does it give away the answer?)
    - [ ]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [ ]  Specific (does the fail message relate to the question?)
- [ ]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [ ]  EOF newlines
- [ ]  Do the comments for Parsons follow the correct convention?
- [ ]  Is the question being tested sufficiently?
    - [ ]  Are there representative cases to test the program?
    - [ ]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [ ]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [ ] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/loops/programs_with_repetition/hard/q7/SinOfTheTimes.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package loops.programs_with_repetition.hard.q7;
+
+/*
+Write a program called SinOfTheTimes that will calculate an approximation of sin(x) from n = 0 until n = 10, with x = 3.14,
```
Comment (Hedgemon4) [inline src/loops/programs_with_repetition/hard/q7/SinOfTheTimes.java:4]: I think we should change this question to specify one of the two numbers, so they can't hardcode it.

Comment (Hedgemon4) [reply src/loops/programs_with_repetition/hard/q7/SinOfTheTimes.java:4]: And then we can vary the other one on the website

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/loops/programs_with_repetition/hard/q7/SinOfTheTimes.java (hunk 1 above)
Comment (Hedgemon4) [reply src/loops/programs_with_repetition/hard/q7/SinOfTheTimes.java:4]: Yeah, but if we change the values, they could still just use something like wolfram alpha to quickly calculate the series, and then just print that out.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Comment by SamamthaH** [comment]
I've changed it over to an input test but in doing so realized there's actually a problem with running this program with java's limited int size. We can't make n greater than 5 because afterwards the factorials get too big for java integers to handle, and we can't approximate higher than 4.4 without the approximation going higher than 1 or lower than -1 for an unknown reason. Because the n is so limited people won't need to use loops, they can just hardcode in 5 lines of the equation so I don't know if this can work as an actual question.

File: src/loops/programs_with_repetition/hard/q7/SinOfTheTimes.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package loops.programs_with_repetition.hard.q7;
+
+/*
+Write a program called SinOfTheTimes that will calculate an approximation of sin(x) from n = 0 until n = 5, with x = 3.14,
```
Comment (Hedgemon4) [inline src/loops/programs_with_repetition/hard/q7/SinOfTheTimes.java:4]: This text is not up to date with the question being changed to have an input.

File: src/loops/programs_with_repetition/hard/q7/SinOfTheTimes.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+using the following series ( n! symbolises factorials, where n! = n (n - 1)( n - 2 )...(1) ):
+x - ( x^3 / 3! ) + ( x^5 / 5! ) - ( x^7 / 7! ) + … + ((-1)^n x^(2n + 1))/( 2n + 1)!
+
+Sample Output:
+	Approximation of sin(3.14) using n = 5 is …
```
Comment (Hedgemon4) [inline src/loops/programs_with_repetition/hard/q7/SinOfTheTimes.java:9]: The sample output is also out of date.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Comment by Hedgemon4** [comment]
> I've changed it over to an input test but in doing so realized there's actually a problem with running this program with java's limited int size. We can't make n greater than 5 because afterwards the factorials get too big for java integers to handle, and we can't approximate higher than 4.4 without the approximation going higher than 1 or lower than -1 for an unknown reason. Because the n is so limited people won't need to use loops, they can just hardcode in 5 lines of the equation so I don't know if this can work as an actual question.

This can be solved by changing the calculation to use double values instead of integer ones.

File: src/loops/programs_with_repetition/hard/q7/SinOfTheTimes.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+/*
+Write a program called SinOfTheTimes that will take an input x and use it to calculate an approximation of sin(x)
+from n = 0 until n = 10 using the following series ( n! symbolises factorials, where n! = n (n - 1)( n - 2 )...(1) ):
+x - ( x^3 / 3! ) + ( x^5 / 5! ) - ( x^7 / 7! ) + … + ((-1)^n x^(2n + 1))/( 2n + 1)! Calculate using doubles due to the size limit of integers.
+
+Sample Output:
+    Enter a number to approximate:
+    3.14
+	Approximation of sin(3.14) using n = 10 is …
```
Comment (Hedgemon4) [inline src/loops/programs_with_repetition/hard/q7/SinOfTheTimes.java:11]:
```suggestion
    Approximation of sin(3.14) using n = 10 is x
```

File: src/loops/programs_with_repetition/hard/q7/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    static Stream<Arguments> inputProvider(){
+        return Stream.of(Arguments.of(2.56, "0.5493554364272203"), Arguments.of(1.07, "0.8772005042746817"), Arguments.of(4.2, "-0.8715757642927082"), Arguments.of(24, "1.052000954450763E9"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    void correctApproximationCalculated(double in, String approx){
+        runWithInput(String.valueOf(in));
+        assertEquals(approx, getItemByName("approx"), "Your approximation is incorrect.");
```
Comment (Hedgemon4) [inline src/loops/programs_with_repetition/hard/q7/MainTest.java:47]:
```suggestion
        assertEquals(approx, getItemByName("approx"), "Your approximation of sin(x) is incorrect.");
```

Comment (Hedgemon4) [inline src/loops/programs_with_repetition/hard/q7/MainTest.java:47]:
```suggestion
        assertEquals(approx, getItemByName("approx"), "Your approximation of sin(x) is incorrect.");
```

File: src/loops/programs_with_repetition/hard/q7/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    public void runMain(){SinOfTheTimes.main(new String[0]);}
+
+    static Stream<Arguments> inputProvider(){
+        return Stream.of(Arguments.of(2.56, "0.5493554364272203"), Arguments.of(1.07, "0.8772005042746817"), Arguments.of(4.2, "-0.8715757642927082"), Arguments.of(24, "1.052000954450763E9"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    void correctApproximationCalculated(double in, String approx){
```
Comment (Hedgemon4) [inline src/loops/programs_with_repetition/hard/q7/MainTest.java:45]:
```suggestion
    void calculatesApproximationCorrectly(double in, String approx){
```

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**